public/components/query_compare/search_result/index.tsx

```diff
@@ -26,77 +32,112 @@ export const SearchResult = ({ http }: SearchResultProps) => {
   const [queryString2, setQueryString2] = useState(DEFAULT_QUERY);
   const [queryResult1, setQueryResult1] = useState<SearchResults>({} as any);
   const [queryResult2, setQueryResult2] = useState<SearchResults>({} as any);
+  const [queryError1, setQueryError1] = useState<QueryError>(initialQueryErrorState);
+  const [queryError2, setQueryError2] = useState<QueryError>(initialQueryErrorState);
   const [searchBarValue, setSearchBarValue] = useState('');
 
   const {
     updateComparedResult1,
     updateComparedResult2,
     selectedIndex1,
     selectedIndex2,
-    setQueryError1,
-    setQueryError2,
   } = useSearchRelevanceContext();
 
   const onClickSearch = () => {
-    const queryErrors = [{ ...initialQueryErrorState }, { ...initialQueryErrorState }];
     const jsonQueries = [{}, {}];
 
-    validateQuery(selectedIndex1, queryString1, queryErrors[0]);
-    jsonQueries[0] = rewriteQuery(searchBarValue, queryString1, queryErrors[0]);
+    validateQuery(selectedIndex1, queryString1, setQueryError1);
+    jsonQueries[0] = rewriteQuery(searchBarValue, queryString1, setQueryError1);
 
-    validateQuery(selectedIndex2, queryString2, queryErrors[1]);
-    jsonQueries[1] = rewriteQuery(searchBarValue, queryString2, queryErrors[1]);
+    validateQuery(selectedIndex2, queryString2, setQueryError2);
+    jsonQueries[1] = rewriteQuery(searchBarValue, queryString2, setQueryError2);
 
-    handleQuery(jsonQueries, queryErrors);
+    handleSearch(jsonQueries);
   };
 
-  const validateQuery = (selectedIndex: string, queryString: string, queryError: QueryError) => {
+  const validateQuery = (
+    selectedIndex: string,
+    queryString: string,
+    setQueryError: React.Dispatch<React.SetStateAction<QueryError>>
+  ) => {
     // Check if select an index
     if (!selectedIndex.length) {
-      queryError.selectIndex = SelectIndexError.unselected;
+      setQueryError((error: QueryError) => ({
+        ...error,
+        selectIndex: SelectIndexError.unselected,
+      }));
     }
 
     // Check if query string is empty
     if (!queryString.length) {
-      queryError.queryString = QueryStringError.empty;
+      setQueryError((error: QueryError) => ({
+        ...error,
+        queryString: QueryStringError.empty,
+        errorRepsonse: {
+          body: '',
+          statusCode: 400,
+        },
+      }));
     }
   };
```

Comment:
Using functional forms of setState makes sense [here](https://github.com/opensearch-project/dashboards-search-relevance/pull/267/files#diff-cabbb4d2260bb522d7fd24de4cde6924fb72ddf93ca92d0fdff83f43630065a2L45-R59) to handle potential state consistency. For validateQuery and rewriteQuery, if it's being refactor to pass state setter instead of the queryError object, the states are updated instantly. Is it a concern since they are helper functions of `onClickSearch`?

Reply:
typo: errorRepsonse -> errorResponse

Reply:
That's a good point. Had a discussion offline, and we concluded that the `validateQuery` helper function does more than what it is supposed to do. Ideally, a validation function should only run validation logic, i.e. it should only return `true` or `false` and nothing else; the setting should be done in another function. [Previously](https://github.com/opensearch-project/dashboards-search-relevance/blob/main/public/components/query_compare/search_result/index.tsx#L53-L63), a variable is passed into `validateQuery` and then is set if it meets conditions. That variable is used later to [set an error state](https://github.com/opensearch-project/dashboards-search-relevance/blob/main/public/components/query_compare/search_result/index.tsx#L80). I believe that the `validateQuery` should be used to return `true` or `false` in the function that sets states to avoid unnecessary stateful functions, and the error state should be set in a stateful function to maintain consistency.